api/key/benchmark_test.go

```diff
@@ -0,0 +1,69 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package key_test
+
+import (
+	"context"
+	"testing"
+
+	"go.opentelemetry.io/otel/api/global"
+	"go.opentelemetry.io/otel/api/key"
+	"go.opentelemetry.io/otel/api/trace"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+)
+
+// Note: The tests below load a real SDK to ensure the compiler isn't optimizing
+// the test based on global analysis limited to the NoopSpan implementation.
+func getSpan() trace.Span {
+	_, sp := global.Tracer("Test").Start(context.Background(), "Span")
+	tr, _ := sdktrace.NewProvider()
+	global.SetTraceProvider(tr)
+	return sp
+}
+
+func BenchmarkMultiNoKeyInference(b *testing.B) {
+	sp := getSpan()
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		sp.SetAttributes(key.Int("Attr", 1))
+	}
+}
+
+func BenchmarkMultiWithKeyInference(b *testing.B) {
+	sp := getSpan()
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		sp.SetAttributes(key.Infer("Attr", 1))
+	}
+}
+
+func BenchmarkSingleWithKeyInferenceValue(b *testing.B) {
+	sp := getSpan()
+
+	b.ReportAllocs()
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		sp.SetAttribute("Attr", 1)
```

Comment:
My 2 cents - this could be `sp.SetAttribute(key.Int("Attr", 1))` as it was before. It is faster and more in line with `SetAttributes` - so the API is more verbose but simpler to understand.

Reply:
Okay, but the point from previous PR was that `SetAttributes` requires slice allocation and that is why faster `SetAttribute` is needed :/

Reply:
Ah, no, that wasn't why I originally introduced SetAttribute, it was solely for brevity. The additional performance benefits or not were a side effect.

Reply:
The motivation in the _original_ SetAttribute was to avoid the cost of a slice. In that thread, you'll see it was argued that we shouldn't support two forms just for a small optimization. This is discussed in #302.

The version that @lizthegrey introduced is more about convenience, see the review for #674, where I wrote how I see it as a good compromise. After doing these benchmarks, it's clear we can have the performance advantage of no allocations (~50ns) for the slight cost (~5ns) of calling `Infer()`, and that use of an `interface{}` in this context does not cause the value to escape to the heap.

Reply:
Thanks, understood.